docs: reorganize content of readme.md

molant opened the pull request:
Fix #1231

molant commented:
@jabfrei can you please take a look at this too?

jabfrei commented on README.md:
```diff
@@ -33,25 +86,169 @@ VS Code will generate a `launch.json` in your project with some default configur
+1. If you have not used exponent on this system before, you will be prompted for an exponent username and password.
+   If you have not created an exponent account, then specifying a new username and password will create one.
+   Note that there is no e-mail associated with the account, and no way to recover a forgotten password.
+1. Once the packager starts, the extension will open a separate tab with QR code to scan from the Exponent app. Once you do so, the Exponent app will connect to the packager and begin running your app.
```
Expo, exponent, and Exponent are mixed here, should be consistent with case

molant replied:
```suggestion
1. If you have not used Exponent on this system before, you will be prompted for an Exponent username and password.
   If you have not created an Exponent account, then specifying a new username and password will create one.
   Note that there is no e-mail associated with the account, and no way to recover a forgotten password.
1. Once the packager starts, the extension will open a separate tab with QR code to scan from the Exponent app. Once you do so, the Exponent app will connect to the packager and begin running your app.
```

jabfrei commented on README.md:
````diff
@@ -33,25 +86,169 @@ VS Code will generate a `launch.json` in your project with some default configur
+}
+```
+
+Full list of configurational parameters for `expo` section in `app.json` may be found on [official Expo documentation page](https://docs.expo.io/versions/latest/workflow/configuration).
````
should this be configuration over configurational?

molant replied:
> should this be configuration over configurational?

configuration sounds better to me but English is not my first language so 🤷‍♂️

```suggestion
Full list of configuration parameters for `expo` section in `app.json` may be found on [official Expo documentation page](https://docs.expo.io/versions/latest/workflow/configuration).
```

jabfrei commented on README.md:
```diff
@@ -33,25 +86,169 @@ VS Code will generate a `launch.json` in your project with some default configur
+
+Full list of configurational parameters for `expo` section in `app.json` may be found on [official Expo documentation page](https://docs.expo.io/versions/latest/workflow/configuration).
+
+For running **pure React Native app**, extension, also, creates and uses `.vscode/exponentIndex.js` which is point to the app entrypoint (`index.js` or `index.android.js` or `index.ios.js`) file.
```
Grammar confusing here, not sure of all the terms here but this is best guess: "Running the **pure React Native app** extension also creates and uses `.vscode/exponentIndex.js`, which points to the app entrypoint (`index.js` or `index.android.js` or `index.ios.js`) file."

molant replied:
```suggestion
For running **pure React Native app**, the extension, creates and uses `.vscode/exponentIndex.js` which points to the app entrypoint (`index.js` or `index.android.js` or `index.ios.js`) file.
```

@SounD120 @RedMickey can you please double check this makes sense?

jabfrei commented on README.md:
```diff
@@ -33,25 +86,169 @@ VS Code will generate a `launch.json` in your project with some default configur
+* Install [ios-deploy](https://www.npmjs.com/package/ios-deploy) `npm install -g ios-deploy`.
+* Have a valid iOS Development certificate installed.
+* In your project's `launch.json` file set `target` to `device`. If you need to specify the exact device to run you can set `target` to `device=<iOS_device_name>` or you can also use `runArguments` property to specify particular device to run on in case of multiple devices connected, e.g. `"runArguments": [ "--device", "My iPhone" ]`
+* Choose **Debug iOS** configuration from the Configuration dropdown and press F5.
```
should Configuration be capitalized?

molant replied:
```suggestion
* Choose the **Debug iOS** option from the "Configuration" dropdown and press F5.
```

@jabfrei what do you think of this?

jabfrei commented on README.md:
```diff
@@ -33,25 +86,169 @@ VS Code will generate a `launch.json` in your project with some default configur
+
+Debugging on an iOS device require following manual steps:
+* Install [ios-deploy](https://www.npmjs.com/package/ios-deploy) `npm install -g ios-deploy`.
+* Have a valid iOS Development certificate installed.
```
Should Development be capitalized here?

molant replied:
```suggestion
* Have a valid iOS development certificate installed.
```

This step was before in another example. We should be consistent with the wording. I believe it was:

> Install a valid iOS development certificate.

I'm OK either way. What do people prefer?

RedMickey replied:
On the [official Apple page](https://developer.apple.com/support/certificates/), they often use the "Developer" with a capital letter, e.g. "Developer ID". So I suppose that the word "Development" could be capitalized.

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
 |`scheme`|A scheme name to be passed to `react-native run-ios`, e.g. `devDebug` to specify `--scheme=devDebug`|`string`|n/a|
 |`productName`|iOS bundle display name e.g. `AwesomeProject` value means that extension will search for `AwesomeProject.app` bundle|`string`|n/a|
```
...means that the extension will...

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
-## Using React Native commands in the Command Palette
+# Customization
+
+The extension can be further customized for other React Native scenarios, these are the most common ones:
```
...scenarios. These are the most common:

jabfrei commented on README.md:
````diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+```
 
-In the Command Palette, type `React Native` and choose a command.
+To lear more about `tasks` in VSCode read [the official documentation](https://code.visualstudio.com/docs/editor/tasks).
````
To learn more...

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+To lear more about `tasks` in VSCode read [the official documentation](https://code.visualstudio.com/docs/editor/tasks).
 
-![React Native commands](images/command-palette.png)
+Visit [generating Signed APK](https://reactnative.dev/docs/signed-apk-android.html) to know more about this subject.
```
... to learn more about...

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+## Specifying custom arguments for `react-native run-*` command
 
-The **Run iOS** command similarly triggers `react-native run-ios` and starts your app in the iOS simulator (iPhone 6).
+For using custom run arguments for `react-native run-<platform>`:
```
Using custom run arguements...

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+
+**NOTE:** You can get the list of installed simulator devices by:
+
+iOS devices(MacOS only):
```
iOS devices (macOS only):

jabfrei commented on README.md:
````diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+adb devices
+```
+
+**NOTE:** If you want run application on iOS devices make sure you have `ios-deploy` installed globally.
````
...want to run an application on iOS devices, make sure you...

jabfrei commented on README.md:
````diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+}
+```
+
+If you change this port, then for iOS device and simulator scenarios you will have to modify the native code files. Instruction [here](https://blog.binoy.io/running-react-native-on-a-different-port-7deb43887cd4).
````
Instructions [here]

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+
+If you change this port, then for iOS device and simulator scenarios you will have to modify the native code files. Instruction [here](https://blog.binoy.io/running-react-native-on-a-different-port-7deb43887cd4).
+
+If you use android, you need to change debug server by:
```
If you use Android, you need to change the debug server by:

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+1. CTRL+M(CMD+M) in the emulator
+2. Go to Dev Settings
+3. Debug server host for device => enter ‘localhost:\<yourPortNumber\>’.
+4. Reload application (double R)
```
Is double R pressing R twice?

molant replied:
Yes

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
-The **Packager** commands allow you to start/stop the [**Metro Bundler**](https://github.com/facebook/metro-bundler) (formerly React Packager).
+# Developing inside a Docker Container
+
+The extension supports [VS Code Remote Development](https://code.visualstudio.com/docs/remote/remote-overview) features on Linux. Please follow the [VS Code official documentation](https://code.visualstudio.com/docs/remote/containers) to setup your environment to use remote development approach.
```
...to use a remote development approach.

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+
+The extension supports [VS Code Remote Development](https://code.visualstudio.com/docs/remote/remote-overview) features on Linux. Please follow the [VS Code official documentation](https://code.visualstudio.com/docs/remote/containers) to setup your environment to use remote development approach.
+
+You can use [official React Native Docker image](https://hub.docker.com/r/reactnativecommunity/react-native-android) provided by [react-native-community](https://github.com/react-native-community/docker-android).
```
...provided by the [react-native-community]...

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+
+You can use [official React Native Docker image](https://hub.docker.com/r/reactnativecommunity/react-native-android) provided by [react-native-community](https://github.com/react-native-community/docker-android).
+
+Here are the steps to run React Native debugging inside Docker Container on a real Android device:
```
...inside a Docker Container...

jabfrei commented on README.md:
````diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+    ```
 
-## Using Expo
+1. Configure your `devcontainer.json` file just about like this: <br> **NOTE**: This is just a sample of configuration, you can modify your `devcontainer.json` file as you need.
````
1. Configure your `devcontainer.json` file as needed. Below is an sample configuration:

jabfrei commented on README.md:
````diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+    ```
 
-## Customization
+1. Open Command Palette and run the following command `Remote-Containers: Open Folder in Container` to reopen your project in container
````
... your project in a container.

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
-## Contributing
+# Contributing
 
 Please see our [contributing guide](CONTRIBUTING.md) for more information
```
... for more information.

jabfrei commented on README.md:
```diff
@@ -71,39 +268,210 @@ See [Setting up debug environment](doc/debugging.md) for more details.
 
-## Customization
+1. Open Command Palette and run the following command `Remote-Containers: Open Folder in Container` to reopen your project in container
+1. Connect your device via USB and start debugging the same way as on local machine
```
...as on a local machine.

molant commented on README.md:
```diff
@@ -3,23 +3,76 @@
+- [Debugging React Native applications](#debugging-react-native-applications)
+  - [Hermes](#hermes)
+  - [iOS devices](#ios-devices)
+  - [Custmom scheme for iOS apps](#custom-scheme-for-ios-apps)
```
It does not have to be a sub-heading of iOS. One is for real devices, the other is for iOS in general.

molant commented on README.md:
```diff
@@ -3,23 +3,76 @@
+  - [TypeScript and Haul](#typescript-and-haul)
+  - [Expo applications](#expo-applications)
+    - [Configuring Expo](#configuring-expo)
+  - [React Native for Windows](#react-native-for-windows)
```
Yeah, I had some mixed feeling about this. On one side we have iOS and Android support out of the box with some particular things for each one, then there's Expo, and finally we have Windows which is a bit more complicated to set up.
I put it also towards the end based on the current popularity. I'll give it a spin see what I come up with.

molant commented on README.md:
```diff
@@ -3,23 +3,76 @@
+  - [Expo applications](#expo-applications)
+    - [Configuring Expo](#configuring-expo)
+  - [React Native for Windows](#react-native-for-windows)
+  - [Debug configuration properties](#debug-configuration-properties)
```
Debugger 👍

molant commented on README.md:
```diff
@@ -33,25 +86,169 @@ VS Code will generate a `launch.json` in your project with some default configur
+
+## TypeScript and Haul
+
+If you use Haul instead of the react-native packager, you must add `sourceMapPathOverrides` to the `launch.json` file.
```
I'm thinking about linking to their website. What do you think?
Also we should probably mention the react-native packager is [Metro](https://facebook.github.io/metro/) 🤔

molant commented:
@jabfrei I think I've commited  all your suggestions and addressed all your comments. Let me know if there's something else that should be modified.

@SounD120 @RedMickey when you have time can you please take a look?

Thanks 🙏

RedMickey commented on README.md:
````diff
@@ -69,41 +269,212 @@ See [Setting up debug environment](doc/debugging.md) for more details.
+    ```
 
-## Using Expo
+1. Configure your `devcontainer.json` file as needed. Below is an sample configuration:
````
```suggestion
1. Configure your `devcontainer.json` file as needed. Below is a sample configuration:
```

RedMickey commented on README.md:
```diff
@@ -2,24 +2,77 @@
+
+The **Run Android** command triggers `react-native run-android` and starts your app for Android.
+
+The **Run iOS** command similarly triggers `react-native run-ios` and starts your app in the iOS simulator (iPhone 6).
```
```suggestion
The **Run iOS** command similarly triggers `react-native run-ios` and starts your app in the iOS simulator (e.g. iPhone 6).
```

RedMickey commented on README.md:
```diff
@@ -33,25 +86,172 @@ VS Code will generate a `launch.json` in your project with some default configur
+
+## Expo applications
+
+To debug a project created using Expo or the create-react-native-app task, you can use embedded support for Expo.
```
```suggestion
To debug a project created using Expo or the `create-react-native-app` task, you can use embedded support for Expo.
```

RedMickey commented on README.md:
```diff
@@ -33,25 +86,172 @@ VS Code will generate a `launch.json` in your project with some default configur
 
-Since version 0.60.2 [React Native supports Hermes engine for Android applications](https://facebook.github.io/react-native/blog/2019/07/17/hermes). Extension provides experimental support for debugging React Native Android applications with Hermes enabled. Please see [Debugging React Native apps with Hermes enabled](doc/debugging.md#debugging-react-native-apps-with-hermes-enabled) for more detai ...
+- Install the [Expo app](https://getexponent.com/) on the target device or emulator
+- Ensure that the react-native-cli is installed globally (`npm install -g react-native-cli`)
```
```suggestion
- Ensure that the `react-native-cli` is installed globally (`npm install -g react-native-cli`)
```

SounD120 commented on README.md:
```diff
@@ -2,24 +2,77 @@
+
+The full list of commands is:
+
+|Name|Description|
```
With that representation that table is quite poor. Command names just duplicate the descriptions. I think it could be useful if we expand descriptions a bit.

molant replied:
I agree, but maybe not a blocker for this PR?
I'll open a separate issue to track this.

molant replied:
I've opened #1246

SounD120 replied:
Of course, I just mentioned this to make sure that we will do it in the future.

SounD120 commented on README.md:
```diff
@@ -33,25 +86,172 @@ VS Code will generate a `launch.json` in your project with some default configur
 
-Since version 0.60.2 [React Native supports Hermes engine for Android applications](https://facebook.github.io/react-native/blog/2019/07/17/hermes). Extension provides experimental support for debugging React Native Android applications with Hermes enabled. Please see [Debugging React Native apps with Hermes enabled](doc/debugging.md#debugging-react-native-apps-with-hermes-enabled) for more detai ...
+- Install the [Expo app](https://getexponent.com/) on the target device or emulator
+- Ensure that the react-native-cli is installed globally (`npm install -g react-native-cli`)
```
```suggestion
- Ensure that the react-native-cli is installed globally (`npm install -g @react-native-community/cli`)
```

molant replied:
I believe this whole section is outdated. Installing globally the RN CLI is not recommended anymore and the steps to get [Expo up and running nowadays seem to be different](https://reactnative.dev/docs/environment-setup):

> Assuming that you have Node 12 LTS or greater installed, you can use npm to install the Expo CLI command line utility:
> `npm install -g expo-cli`
> Then run the following commands to create a new React Native project called "AwesomeProject":
> ```
> expo init AwesomeProject`
> cd AwesomeProject
> npm start # you can also use: expo start
> ```
> This will start a development server for you.

@SounD120 @RedMickey can you please confirm and tell me if there's any other requirements than the ones in the docs?

SounD120 replied:
@molant Actually we might even don't need `expo-cli` at all for running Expo debugging scenarios since we install [`@expo/xdl`](https://www.npmjs.com/package/@expo/xdl) package inside the extension which is doing all the communication with Expo services. But it might be useful to init fresh Expo apps and to work with some of the extension commands. With respect to the React Native CLI - since [v0.13.0](https://github.com/microsoft/vscode-react-native/blob/master/CHANGELOG.md#0130) we can work with the locally installed CLI so the global one is not required.
@RedMickey correct me if I missed something.

RedMickey replied:
Yep, It's possible to debug an Expo app without having `expo-cli` installed. I tried to do this and was able to debug an Expo app without problems.
Also now we don't use global react-native CLI.

molant replied:
@RedMickey @SounD120 do you mind suggesting new instructions for this section?

Thanks!

SounD120 replied:
Yes, it is better to update them

molant replied:
@RedMickey @SounD120 what would be the new steps for this? I believe it's the last comment to get this merged 😊

SounD120 replied:
I think we'll finally merge this and do all enhancements in another PRs. It was here for too long because of problems with CI tests, but now they gone, and I think we are good to go :)

molant replied:
SGTM!
Will you take care of merging? Something else you need me to do?

SounD120 replied:
I am waiting for the tests now, once they are passed I'll merge this.

RedMickey commented on README.md:
```diff
@@ -33,25 +86,172 @@ VS Code will generate a `launch.json` in your project with some default configur
+1. Create a debug configuration (as described in [Debugging React Native applications](#debugging-react-native-applications)), select `Debug in Exponent` in the debug drop-down menu, and start debugging
+1. Wait while some dependencies are configured - the extension will install `xde` and `@expo/ngrok` when this feature is used for the first time.
+1. If you have not used Exponent on this system before, you will be prompted for an Exponent username and password.
+   If you have not created an Exponent account, then specifying a new username and password will create one.
```
```suggestion
   Exponent account allows you to use Expo cloud services. More info about how it works is available [here](https://docs.expo.io/versions/latest/workflow/how-expo-works/).
   If you have not created an Exponent account, then specifying a new username and password will create one.
```

SounD120 commented on README.md:
```diff
@@ -2,24 +2,77 @@
 
 [![Build status](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_apis/build/status/vscode-react-native%20%5Bmaster%5D)](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_build/latest?definitionId=2)
```
```suggestion
[![Build status](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_apis/build/status/vscode-react-native%20%5Bmaster%5D)](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_build/latest?definitionId=60)
```

RedMickey commented on README.md:
```diff
@@ -33,25 +86,172 @@ VS Code will generate a `launch.json` in your project with some default configur
+1. Wait while some dependencies are configured - the extension will install `xde` and `@expo/ngrok` when this feature is used for the first time.
+1. If you have not used Exponent on this system before, you will be prompted for an Exponent username and password.
+   If you have not created an Exponent account, then specifying a new username and password will create one.
+   Note that there is no e-mail associated with the account, and no way to recover a forgotten password.
```
```suggestion
   Note that there is no e-mail associated with the account, and no way to recover a forgotten password.
   If you don't want to create an Exponent account, you can specify `expoHostType` parameter in your debug configuration to make Expo work locally (via LAN or on localhost).
```
